### ISSUE TITLE
How to run Aspire on multiple computers instead of running Aspire on Azure

### ISSUE THREAD

**Issue by cysnet** [dotnet/aspire#831]
How to run Aspire on multiple computers instead of running Aspire on Azure?

**Comment by cysnet** [comment]
Does Aspire only run in Azure?

**Comment by DamianEdwards** [comment]
Aspire doesn't include a runtime control plane or hosting mechanism for outside of the developer scenario. It's intended the app is deployed to something like kubernetes or a container-based PaaS. Ultimately the apps are standard .NET projects that you can deploy and configure in any of the supported ways.

**Comment by HClausing** [comment]
I'm doing some tests here with the starter template, and it's working fine just with Kestrel.

**Comment by cysnet** [comment]
@DamianEdwards Is there a document to describe how to deploy aspire applications to k8s?

**Comment by davidfowl** [comment]
There is no automated deployment to k8s support in aspire preview1. Deploying to k8s is manual right now. We plan to investigate this in the future.

**Comment by cysnet** [comment]
@davidfowl how to deploy aspire applications to k8s manually? create deployments for every aspire application?

**Comment by cysnet** [comment]
@davidfowl Will multi-virtual machine deployment be supported in the future instead of k8s?

**Comment by davidfowl** [comment]
> @davidfowl Will multi-virtual machine deployment be supported in the future instead of k8s?

No, never. We're not rebuilding kubernetes. The orchestration here is optimized for development scenarios. If you want anything on multiple machines, then use k8s.

**Comment by cysnet** [comment]
@davidfowl When is it expected that there will be documentation on how to deploy through k8s? And before this, could I only deploy to Azure?

**Comment by davidfowl** [comment]
Support for automatically generating and deploying to k8s? Months away. Maybe 2 (this experience needs to be designed). Yes, you can deploy to azure using AZD. Aspire itself cannot deploy to azure, the azure tools understand aspire manifests.

If you want to help design the k8s experience follow and contribute to this issue https://github.com/dotnet/aspire/issues/830

**Comment by HClausing** [comment]
> > @davidfowl Will multi-virtual machine deployment be supported in the future instead of k8s?
> 
> No, never. We're not rebuilding Kubernetes. The orchestration here is optimized for development scenarios. If you want anything on multiple machines, then use k8s.

Hi @davidfowl , I'm not going against the plans. But not only me, but many users from communities that I participate liked a lot what we've seen on .NET Conf about the .NET Aspire. I guess it so much powerfull for just to stay only on development scenarios. It's very easy to setup and run just using Kestrel + YARP (when we don't need to use K8S neither Docker).

We'll very happy we could see .NET Aspire to running also on production scenarios.

**Comment by davidfowl** [comment]
I don't think there's anything wrong with your desire to want this experience for deployments to non-cloud environments, but the reason we target the cloud and things like k8s for deployment is because there's a standard set of APIs for interacting with and configuring the compute, networking and anything required to make these systems run.

If you want to deploy to another environment then you can use aspire's extensible system but you have to build a tool that understands your target environment. I can help guide you to build it, but it's not something we are building as part of aspire.

**Comment by HClausing** [comment]
> I don't think there's anything wrong with your desire to want this experience for deployments to non-cloud environments, but the reason we target the cloud and things like k8s for deployment is because there's a standard set of APIs for interacting with and configuring the compute, networking and anything required to make these systems run.
> 
> If you want to deploy to another environment then you can use aspire's extensible system but you have to build a tool that understands your target environment. I can help guide you to build it, but it's not something we are building as part of aspire.

Cool, thank you so much for the answer!

What I'll try to do is to deploy to a server using only Kestrel + Aspire + YARP. It's alread works fine, but I've noted that Publish option on VS Preview 17.9 is compiling the projects looks like the Debug configuration, and it doesn't  copy the compiled files to desired target folder (just for the \bin\Release).

After it, I can immagine for the first time an extension that can allow to restart some service from the list on dashboard, but just for authenticated admins (role).

Do you believe it can be possible? For sure, not from official Aspire package, since it's not on roadmap, but into another extension.

**Comment by davidfowl** [comment]
Are you running your applications in containers or just as executables on the machine? 

I find these discussions to be more useful with diagrams:

![image](https://github.com/dotnet/aspire/assets/95136/4cf0f925-54f3-43aa-8080-5900567a6533)

What features of aspire do you want to persist in your production environment?
- The dashboard?
- The orchestrator (which is only usable in development)? If yes, what are you using today to run your applications in prod? 
- What else?

**Comment by HClausing** [comment]
> Are you running your applications in containers or just as executables on the machine?

Actually I pretend to run on the prodution environment without containers. I like the ideia of Aspire starting every depedant project or executable by itself and showing their status into a deshboard (And maybe on future ofering a restart option on dashboard or function .Restart(...); ).

> What features of aspire do you want to persist in your production environment?
> 
> * The dashboard 👍🏾
> * service discovery 👍🏾
> * The orchestrator (which is only usable in development)? If yes, what are you using today to run your applications in prod?
> * What else?

* The dashboard:  It's good to check the health/logs from every dependant project/executable (services / microservices);
* Service discovery: Service discovery: I didn't kwon this feature until aspire, and I liked it alot;
* The orchestrator: Maybe I misunderstood this feature, but nowadays I'm running every executable from required services by scheduled tasks on OS (triggering they from OS startup - not requires an user to log on). This feature will simplify it into a single scheduled task instead many. It can allow on-premises deployment, for example, without docker and k8s.

Based on diagram, without Aspire, and not using Docker, I should configure four scheduled tasks on OS, and sometimes one or more can exit and the related services will also fail, and we only discover by the Task Manager (if Windows Server) or Event Viewer. The .NET Aspire on this preview state already helps a lot.

**Comment by samsp-msft** [comment]
Maybe we should be looking at how to enable the projects for Aspire be more easily run as a windows service or linux daemon for non-cloud deployment scenarios? As long as the apps can be F5'd while in VS, the eventual hosting model shouldn't matter.
